### PR TITLE
Use custom TimeSpanParser

### LINKED ISSUE #154: Negative sign in /remind offset is ignored
### Description

When specifying a negative duration for the /remind offset, the negative sign is ignored and the equivalent size positive duration applies instead 

### Reproduction Steps

1. Run `/remind` with a negative offset specified (e.g. `/remind in:-5m message:test`) 

### Expected behavior

The bot gives an error and doesn't create a reminder
OR
The bot warns the user about autocorrection and still creates a reminder like it does now

### Actual behavior

The bot silently uses an equivalent size positive duration of the specified negative duration

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by mctaylors**
Closes #154

### CODE REVIEW

**Review by neroduckale — APPROVED** [review]
nice fix

**Review by neroduckale — APPROVED** [review]
re-approve, nice code

File: locale/Messages.resx [hunk 1]
```diff
@@ -585,4 +585,10 @@
   <data name="ButtonReportIssue" xml:space="preserve">
       <value>Report an issue</value>
   </data>
+  <data name="ReminderInvalidTimeSpan" xml:space="preserve">
+      <value>Incorrect reminder send time specified.</value>
+  </data>
+  <data name="ReminderNegativeTimeSpan" xml:space="preserve">
+      <value>Negative reminder send time specified.</value>
+  </data>
```
Comment (Octol1ttle) [inline locale/Messages.resx:593]: I would only use `InvalidTimeSpan` here. Negative TimeSpans do not make sense in any context

File: src/Commands/RemindCommandGroup.cs [hunk 2]
```diff
@@ -74,10 +75,12 @@ public async Task<Result> ExecuteListReminderAsync()
         var data = await _guildData.GetData(guildId, CancellationToken);
         Messages.Culture = GuildSettings.Language.Get(data.Settings);
 
-        return await ListRemindersAsync(data.GetOrCreateMemberData(executorId), guildId, executor, bot, CancellationToken);
+        return await ListRemindersAsync(data.GetOrCreateMemberData(executorId), guildId, executor, bot,
```
Comment (Octol1ttle) [inline src/Commands/RemindCommandGroup.cs:78]: This pull request is *filled* with unnecessary formatting changes.

File: src/Commands/RemindCommandGroup.cs [hunk 3]
```diff
@@ -120,7 +125,7 @@ private Task<Result> ListRemindersAsync(MemberData data, Snowflake guildId, IUse
     [UsedImplicitly]
     public async Task<Result> ExecuteReminderAsync(
         [Description("After what period of time mention the reminder")]
-        TimeSpan @in,
+        string @in,
```
Comment (Octol1ttle) [inline src/Commands/RemindCommandGroup.cs:128]: Now that the type is gone, the parameter name is confusing.

File: src/Commands/RemindCommandGroup.cs [hunk 4]
```diff
@@ -138,15 +149,38 @@ private Task<Result> ListRemindersAsync(MemberData data, Snowflake guildId, IUse
         var data = await _guildData.GetData(guildId, CancellationToken);
         Messages.Culture = GuildSettings.Language.Get(data.Settings);
 
-        return await AddReminderAsync(@in, text, data, channelId, executor, CancellationToken);
+        return await AddReminderAsync(@in, text, data, channelId, bot, executor, CancellationToken);
     }
 
-    private async Task<Result> AddReminderAsync(TimeSpan @in, string text, GuildData data,
-        Snowflake channelId, IUser executor, CancellationToken ct = default)
+    private async Task<Result> AddReminderAsync(string strIn, string text, GuildData data,
```
Comment (Octol1ttle) [inline src/Commands/RemindCommandGroup.cs:155]: Confusing parameter name

File: src/Commands/RemindCommandGroup.cs [hunk 5]
```diff
@@ -138,15 +149,38 @@ private Task<Result> ListRemindersAsync(MemberData data, Snowflake guildId, IUse
...
+            var failedEmbed = new EmbedBuilder()
+                .WithSmallTitle(Messages.ReminderNegativeTimeSpan, bot)
+                .WithColour(ColorsList.Red)
+                .Build();
+
+            return await _feedback.SendContextualEmbedResultAsync(failedEmbed, ct: ct);
+        }
+
+        var parser = new TimeSpanParser();
+        var parseResult = await parser.TryParseAsync(strIn, ct);
```
Comment (Octol1ttle) [inline src/Commands/RemindCommandGroup.cs:169]: If we are gonna stop using automatic parsing, might as well use a custom parser that can accept cyrillic letters as units (like the one used in Discord.NET edition)

**Review by Octol1ttle — CHANGES_REQUESTED** [review]
1) This does not close #220 because the core problem is not fixed (being that parsing errors are not relayed to the end user)
2) Even if we ignore that, the fix provided only applies to `/remind` and not other commands that accept a TimeSpan.

File: src/Commands/RemindCommandGroup.cs (hunk 3 above)
Comment (mctaylors) [reply src/Commands/RemindCommandGroup.cs:128]: Parameter name is also showing when using command. Renaming it will probably affect UX.

Comment (Octol1ttle) [reply src/Commands/RemindCommandGroup.cs:128]: Good point. In that case, you can use the `[Option]` attribute to change the name that will display to the end user.

File: src/Commands/RemindCommandGroup.cs (hunk 2 above)
Comment (mctaylors) [reply src/Commands/RemindCommandGroup.cs:78]: > It's better to reformat code only manually and only when you feel "this code is hard to read and I know how to make it better"

this code is hard to read and but now I don't know how to make it better

File: src/Commands/RemindCommandGroup.cs [hunk 6]
```diff
@@ -120,7 +121,8 @@ private Task<Result> ListRemindersAsync(MemberData data, Snowflake guildId, IUse
     [UsedImplicitly]
     public async Task<Result> ExecuteReminderAsync(
         [Description("After what period of time mention the reminder")]
-        TimeSpan @in,
+        [Option("in")]
+        string stringIn,
```
Comment (Octol1ttle) [inline src/Commands/RemindCommandGroup.cs:125]: confusing parameter name

File: src/Commands/RemindCommandGroup.cs [hunk 7]
```diff
@@ -138,14 +146,30 @@ private Task<Result> ListRemindersAsync(MemberData data, Snowflake guildId, IUse
...
+
+        if (timeSpanIn == TimeSpan.Zero)
+        {
+            var failedEmbed = new EmbedBuilder()
+                .WithSmallTitle(Messages.InvalidTimeSpan, bot)
+                .WithColour(ColorsList.Red)
+                .Build();
+
+            return await _feedback.SendContextualEmbedResultAsync(failedEmbed, ct: ct);
+        }
```
Comment (Octol1ttle) [inline src/Commands/RemindCommandGroup.cs:169]: Parsing and input validation should be done in the preparation step

File: src/Parsers/TimeSpanParser.cs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+///     If parse wasn't successful, <see cref="TimeSpanParser"/> will return <see cref="TimeSpan.Zero"/>.
+/// </remarks>
+[PublicAPI]
+public partial class TimeSpanParser : AbstractTypeParser<TimeSpan>
+{
+    private static readonly Regex Pattern = ParseRegex();
+
+    public static Result<TimeSpan> TryParse(string timeSpanString, CancellationToken ct = default)
+    {
+        var timeSpan = TimeSpan.Zero;
```
Comment (Octol1ttle) [inline src/Parsers/TimeSpanParser.cs:25]: Use `TimeSpan.Zero` explicitly

File: src/Parsers/TimeSpanParser.cs [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        timeSpanString = timeSpanString.Trim();
```
Comment (Octol1ttle) [inline src/Parsers/TimeSpanParser.cs:26]: Is this necessary? Doesn't Discord trim strings before sending them as slash command inputs?

File: src/Parsers/TimeSpanParser.cs [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+                     .Cast<Group>()
+                     .Where(g => g.Success)
+                     .Skip(1)
+                     .Select(g => (g.Name, g.Value))))
+        {
+            foreach ((var key, var groupValue) in groups)
+            {
+                return !double.TryParse(groupValue, out var parsedGroupValue)
+                    ? timeSpan
+                    : ParseFromRegex(timeSpan, key, groupValue, parsedGroupValue);
```
Comment (Octol1ttle) [inline src/Parsers/TimeSpanParser.cs:54]: Invert condition

File: src/Parsers/TimeSpanParser.cs [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+/// </returns>
+/// <remarks>
+///     If parse wasn't successful, <see cref="TimeSpanParser"/> will return <see cref="TimeSpan.Zero"/>.
+/// </remarks>
+[PublicAPI]
+public partial class TimeSpanParser : AbstractTypeParser<TimeSpan>
+{
+    private static readonly Regex Pattern = ParseRegex();
+
+    public static Result<TimeSpan> TryParse(string timeSpanString, CancellationToken ct = default)
```
Comment (Octol1ttle) [inline src/Parsers/TimeSpanParser.cs:23]: The method returns a `Result`, but no errors are returned

File: locale/Messages.ru.resx [hunk 12]
```diff
@@ -585,6 +585,9 @@
   <data name="ButtonReportIssue" xml:space="preserve">
       <value>Сообщить о проблеме</value>
   </data>
+  <data name="InvalidTimeSpan" xml:space="preserve">
+      <value>Указано некорректное время.</value>
```
Comment (Octol1ttle) [inline locale/Messages.ru.resx:589]:
```suggestion
      <value>Неправильно указано время!</value>
```

File: locale/Messages.resx [hunk 13]
```diff
@@ -585,6 +585,9 @@
   <data name="ButtonReportIssue" xml:space="preserve">
       <value>Report an issue</value>
   </data>
+  <data name="InvalidTimeSpan" xml:space="preserve">
+      <value>Incorrect time specified.</value>
```
Comment (Octol1ttle) [inline locale/Messages.resx:589]:
```suggestion
      <value>Time specified incorrectly!</value>
```

**Review by Octol1ttle — CHANGES_REQUESTED** [review]

File: src/Parsers/TimeSpanParser.cs (hunk 9 above)
Comment (mctaylors) [reply src/Parsers/TimeSpanParser.cs:26]: > Doesn't Discord trim strings before sending them as slash command inputs?
only at the beginning

Comment (Octol1ttle) [reply src/Parsers/TimeSpanParser.cs:26]: in that case, strings should come trimmed, the parser shouldn't need to do extra work

Comment (mctaylors) [reply src/Parsers/TimeSpanParser.cs:26]: but what if user will enter `5 min` string?

Comment (Octol1ttle) [reply src/Parsers/TimeSpanParser.cs:26]: what is the issue here?

Comment (mctaylors) [reply src/Parsers/TimeSpanParser.cs:26]: strings come without spaces only *at the beginning and at the end*, between actual characters spaces will remain. won't the parser get confused if there is a space between the number and the unit (?) type?

Comment (Octol1ttle) [reply src/Parsers/TimeSpanParser.cs:26]: > won't the parser get confused if there is a space between the number and the unit (?) type?
if that's the case, then the current version of your code shouldn't even work.

File: src/Commands/RemindCommandGroup.cs [hunk 14]
```diff
@@ -17,6 +17,7 @@
 using Remora.Discord.Extensions.Formatting;
 using Remora.Rest.Core;
 using Remora.Results;
+using TimeSpanParser = Octobot.Parsers.TimeSpanParser;
```
Comment (Octol1ttle) [inline src/Commands/RemindCommandGroup.cs:20]: use a normal `using`

File: src/Parsers/TimeSpanParser.cs [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+    ///     Parsed <see cref="TimeSpan"/>.
+    /// </returns>
+    /// <remarks>
+    ///     If parse wasn't successful, <see cref="TimeSpanParser"/> will return <see cref="TimeSpan.Zero"/>.
+    /// </remarks>
+    public static TimeSpan TryParse(string timeSpanString, CancellationToken ct = default)
+    {
+        if (timeSpanString.StartsWith('-'))
+        {
+            return TimeSpan.Zero;
```
Comment (Octol1ttle) [inline src/Parsers/TimeSpanParser.cs:29]: This should return an error instead of Zero

**Review by Octol1ttle — CHANGES_REQUESTED** [review]

**Review by Octol1ttle — CHANGES_REQUESTED** [review]

File: src/Parsers/TimeSpanParser.cs [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+                     .Where(g => g.Success)
+                     .Skip(1)
+                     .Select(g => (g.Name, g.Value))))
+        {
+            foreach ((var key, var groupValue) in groups)
+            {
+                if (!double.TryParse(groupValue, out var parsedGroupValue) ||
+                    !int.TryParse(groupValue, out var parsedIntegerValue))
+                {
+                    return TimeSpan.Zero;
```
Comment (Octol1ttle) [inline src/Parsers/TimeSpanParser.cs:62]: Once again you are returning `Zero` instead of an error.

File: src/Parsers/TimeSpanParser.cs [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+
+    /// <summary>
+    ///     Parses <see cref="TimeSpan"/> from <see cref="string"/>.
+    /// </summary>
+    /// <returns>
+    ///     Parsed <see cref="TimeSpan"/>.
+    /// </returns>
+    /// <remarks>
+    ///     If parse wasn't successful, <see cref="TimeSpanParser"/> will return <see cref="ArgumentInvalidError"/>.
+    /// </remarks>
```
Comment (Octol1ttle) [inline src/Parsers/TimeSpanParser.cs:25]:
```suggestion
    /// <summary>
    ///     Parses a <see cref="TimeSpan"/> from the <paramref name="timeSpanString"/>.
    /// </summary>
    /// <returns>
    ///     The parsed <see cref="TimeSpan"/>, or <see cref="ArgumentInvalidError"/> if parsing failed.
    /// </returns>
```

File: src/Commands/RemindCommandGroup.cs [hunk 18]
```diff
@@ -120,7 +121,8 @@ private Task<Result> ListRemindersAsync(MemberData data, Snowflake guildId, IUse
     [UsedImplicitly]
     public async Task<Result> ExecuteReminderAsync(
         [Description("After what period of time mention the reminder")]
-        TimeSpan @in,
+        [Option("in")]
+        string stringTimeSpan,
```
Comment (Octol1ttle) [inline src/Commands/RemindCommandGroup.cs:125]: The parameter is called `stringTimeSpan` here, but `timeSpanString` in the parser.

**Review by Octol1ttle — CHANGES_REQUESTED** [review]

File: src/Parsers/TimeSpanParser.cs (hunk 16 above)
Comment (mctaylors) [reply src/Parsers/TimeSpanParser.cs:62]: wait how did I missed that...

File: src/Parsers/TimeSpanParser.cs [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+
+        if (TimeSpan.TryParse(timeSpanString, DateTimeFormatInfo.InvariantInfo, out var parsedTimeSpan))
+        {
+            return parsedTimeSpan;
+        }
+
+        var matches = ParseRegex().Matches(timeSpanString);
+        return matches.Count is 0
+            ? new ArgumentInvalidError(nameof(timeSpanString), "The regex did not produce any matches.")
+            : ParseFromRegex(matches);
```
Comment (Octol1ttle) [inline src/Parsers/TimeSpanParser.cs:38]: Invert

File: src/Parsers/TimeSpanParser.cs [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+        foreach (var groups in matches.Select(match => match.Groups
+                     .Cast<Group>()
+                     .Where(g => g.Success)
+                     .Skip(1)
+                     .Select(g => (g.Name, g.Value))))
+        {
+            foreach ((var key, var groupValue) in groups)
+            {
+                if (!double.TryParse(groupValue, out var parsedGroupValue) ||
+                    !int.TryParse(groupValue, out var parsedIntegerValue))
```
Comment (Octol1ttle) [inline src/Parsers/TimeSpanParser.cs:54]: If the intent is to support `double`s (like `3.5d`), then this code won't work. The input value has to be an integer as per the second condition.
If that is not the intent, the first condition can be omitted (as `int`s are implicitly casted to `double`s when needed)

File: src/Parsers/TimeSpanParser.cs [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+                {
+                    return new ArgumentInvalidError(nameof(groupValue), "The input value encountered a parsing error.");
```
Comment (Octol1ttle) [inline src/Parsers/TimeSpanParser.cs:56]:
```suggestion
                    return new ArgumentInvalidError(nameof(groupValue), "The input value was not an integer.");
```

**Review by Octol1ttle — CHANGES_REQUESTED** [review]

File: src/Parsers/TimeSpanParser.cs (hunk 19 above)
Comment (mctaylors) [reply src/Parsers/TimeSpanParser.cs:38]: I cannot, see https://github.com/LabsDevelopment/Octobot/actions/runs/7370013830/job/20055931614

Comment (Octol1ttle) [reply src/Parsers/TimeSpanParser.cs:38]: *sigh*
```suggestion
        return matches.Count > 0
            ? ParseFromRegex(matches)
            : new ArgumentInvalidError(nameof(timeSpanString), "The regex did not produce any matches.");
```

File: src/Parsers/TimeSpanParser.cs [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+                    "Seconds" => TimeSpan.FromSeconds(parsedIntegerValue),
+                    _ => throw new ArgumentOutOfRangeException(key)
+                };
+            }
+        }
+
+        return timeSpan;
+    }
+
+    [GeneratedRegex("(?<Years>\\d+(?=y|л|г))|(?<Months>\\d+(?=mo|мес))|(?<Weeks>\\d+(?=w|н|нед))|(?<Days>\\d+(?=d|дн))|(?<Hours>\\d+(?=h|ч))|(?<Minutes>\\d+(?=m|min|мин|м))|(?<Seconds>\\d+(?=s|sec|с|сек))")]
```
Comment (Octol1ttle) [inline src/Parsers/TimeSpanParser.cs:76]: Doesn't look like it's possible to use `д` for days

**Review by Octol1ttle — CHANGES_REQUESTED** [review]
The issues highlighted [in my original review](https://github.com/LabsDevelopment/Octobot/pull/223#pullrequestreview-1786032829) haven't been resolved.

**Review by Octol1ttle — APPROVED** [review]

**Pull request merged**